File: dart/trapping-rain-water.dart
class Solution {
  int trap(List<int> height) {
    // highest point on the left of each point
    // height [0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1] ->
    // maxLeft[0, 0, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3]
    var maxLeft = List<int>.filled(height.length, 0);
    for (int i = 1; i < maxLeft.length; i++) {
      maxLeft[i] = max(maxLeft[i - 1], height[i - 1]);
    }
    var maxRight = List<int>.filled(height.length, 0);
    for (int i = maxRight.length - 2; i >= 0; i--) {
      maxRight[i] = max(maxRight[i + 1], height[i + 1]);
    }

    int trappedWater = 0;
    for (int i = 0; i < height.length; i++) {
      // for each point we find amount of water above it
      trappedWater += max(0, min(maxLeft[i], maxRight[i]) - height[i]);
    }
    return trappedWater;
  }
}

File: go/trapping-rain-water.go
func trap(height []int) int {
	highest_on_left := make([]int, len(height))
	for i := 1; i < len(height); i++ {
		if height[i - 1] > highest_on_left[i - 1] {
			highest_on_left[i] = height[i - 1]
		} else {
			highest_on_left[i] = highest_on_left[i - 1]
		}
	}

	highest_on_right := make([]int, len(height))
	for i := len(height) - 2; i >= 0; i-- {
		if height[i + 1] > highest_on_right[i + 1] {
			highest_on_right[i] = height[i + 1]
		} else {
			highest_on_right[i] = highest_on_right[i + 1]
		}
	}

	contained_water := 0
	for i := 0; i < len(height); i++ {
		contained_water += max(
			min(highest_on_right[i], highest_on_left[i]) - height[i],
			0,
		)
	}
	return contained_water
}
